### PR TITLE
feat(contrib/googleapis/go-genai): add Google GenAI Go SDK integration

### PR DESCRIPTION
**Pull request by billyac**
## Summary

Adds Datadog LLM Observability tracing for [`google.golang.org/genai`](https://github.com/googleapis/go-genai) at `contrib/googleapis/go-genai/v2`.

API matches the existing LLMObs contrib precedent (`mark3labs/mcp-go`, `modelcontextprotocol/go-sdk`): the user calls `genaitrace.WrapClient(*genai.Client) *Client` and uses the returned client in place of the original. No `orchestrion.yml` for now — same as the other LLMObs contribs.

**Instrumented methods**
- `Models.GenerateContent`, `Models.GenerateContentStream` (chunks accumulated for output text + final usage)
- `Models.EmbedContent`
- `Chats.Create` → traced `*Chat.SendMessage` / `*Chat.SendMessageStream`

**Span data**
- Model name, model provider (derived from backend: `google` / `google_vertexai` / `google_enterprise`)
- Generation metadata: temperature, top_p, top_k, max_output_tokens, stop_sequences
- Metrics: `input_tokens`, `output_tokens`, `total_tokens`, `cache_read_input_tokens`, `reasoning_output_tokens`

**Repo wiring**
- `instrumentation/packages.go`: `PackageGoogleAPIsGoGenAI`
- `ddtrace/tracer/option.go` `contribIntegrations`: `google.golang.org/genai`
- `internal/env/supported_configurations.{json,gen.go}`: `DD_TRACE_GOOGLE_GENAI_ANALYTICS_ENABLED`
- `internal/stacktrace/contribs_generated.go` (regenerated)
- `go.work`

## Test plan

- [x] `go test ./contrib/googleapis/go-genai/...` — five tests covering generate, generate stream (SSE), embed, chat with prior history (two consecutive sends to verify history-snapshot ordering), and error path, all running against an `httptest` mock server.
- [x] `go test ./instrumentation/...` — `TestInstrumentation_AnalyticsRate/googleapis/go-genai` passes.
- [x] `go test ./ddtrace/tracer/ -run TestAgentIntegration` — passes after bumping default integration count 57 → 58.
- [x] `go build ./...` from repo root.
- [ ] Real Gemini API smoke test (requires API key) — not run in CI.

## Review follow-ups

- **Chat history snapshot** (Codex P2 on `chats.go`): `genai.Chat.SendMessage` appends the current turn to `History()` on success, so reading history after the call duplicated the user message and leaked the assistant reply into the recorded input. Fixed by snapshotting history before the call in a new `snapshotInput` helper; `TestChatSendMessage` now performs two consecutive sends and asserts each span's input contains the new user turn exactly once and no assistant content from the same turn.
- **Supported configuration metadata** (Codex P2 on `supported_configurations.json`): replaced the `FIX_ME` placeholders for `DD_TRACE_GOOGLE_GENAI_ANALYTICS_ENABLED` with `boolean` / default `false`, matching every other `DD_TRACE_*_ANALYTICS_ENABLED` key.

## Outstanding

- `dd-gitlab/validate_supported_configurations_v2_local_file` requires a Datadog maintainer to register `DD_TRACE_GOOGLE_GENAI_ANALYTICS_ENABLED` in the internal configuration registry (per CONTRIBUTING.md). I can't do this from outside.

Claude session: `10d83514-c53a-4629-b0b1-ccde7bb21f6b`
Resume: `claude --resume 10d83514-c53a-4629-b0b1-ccde7bb21f6b`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: contrib/googleapis/go-genai/chats.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+// captured on the span includes the existing chat history plus the new
+// user message.
+func (c *Chat) SendMessage(ctx context.Context, parts ...genai.Part) (*genai.GenerateContentResponse, error) {
+	if c == nil || c.chat == nil {
+		return nil, errNilClient
+	}
+	span, ctx := startLLMSpan(ctx, "genai.chat.send_message", c.model, c.provider)
+
+	resp, err := c.chat.SendMessage(ctx, parts...)
+	finishLLMSpan(span, c.inputForSpan(parts), c.config, resp, err)
```
[the remark this answers is not in the document]
Comment (billyac) [reply contrib/googleapis/go-genai/chats.go]: Good catch — fixed in dcb3fa84a. Snapshotted history into a local before calling SendMessage/SendMessageStream (new snapshotInput helper). Extended TestChatSendMessage to two consecutive sends: each span's input now contains the new user turn exactly once and the assistant reply for that turn doesn't appear in its own input.

File: internal/env/supported_configurations.json [hunk 2]
```diff
@@ -1218,6 +1218,13 @@
         "default": "false"
       }
     ],
+    "DD_TRACE_GOOGLE_GENAI_ANALYTICS_ENABLED": [
+      {
+        "implementation": "A",
+        "type": "FIX_ME",
+        "default": "FIX_ME"
```
[the remark this answers is not in the document]
Comment (billyac) [reply internal/env/supported_configurations.json]: Fixed in dcb3fa84a — set type="boolean", default="false" to match the other DD_TRACE_*_ANALYTICS_ENABLED keys.

File: instrumentation/packages.go [hunk 3]
```diff
@@ -79,6 +79,7 @@ const (
 	PackageUptraceBun                Package = "uptrace/bun"
 	PackageLogSlog                   Package = "log/slog"
 	PackageModelContextProtocolGoSDK Package = "modelcontextprotocol/go-sdk"
+	PackageGoogleAPIsGoGenAI         Package = "googleapis/go-genai"
```
Comment (rarguelloF) [inline instrumentation/packages.go]:
```suggestion
	PackageGoogleAPIsGoGenAI         Package = "google.golang.org/genai"
```

File: contrib/google.golang.org/genai/chats.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (rarguelloF) [inline contrib/google.golang.org/genai/chats.go:1]: the contrib path should match the traced package path

(in this case it should live under `contrib/google.golang.org/genai`)

File: contrib/googleapis/go-genai/client.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	// Models wraps the underlying *genai.Client.Models with tracing.
+	Models *Models
+	// Chats wraps the underlying *genai.Client.Chats with tracing.
+	Chats *Chats
+}
+
+// WrapClient returns a tracing wrapper around the given *genai.Client.
+// The returned *Client exposes wrapper Models and Chats with the same
+// method signatures as the upstream SDK.
+func WrapClient(c *genai.Client) *Client {
```
Comment (rarguelloF) [inline contrib/googleapis/go-genai/client.go]: usually we accept here a variadic `opts ...Option` argument in case we want to provide configuration for the integration. Even if we don't have any option in mind right now, its worth it to add it to avoid introducing breaking changes in the api if we need it in the future.

**Review by rarguelloF — CHANGES_REQUESTED** [review]

File: instrumentation/packages.go (hunk 3 above)
Comment (billyac) [reply instrumentation/packages.go]: Applied in 0e44ff919. Also updated the naming registration and matching MCP-style tests still pass.

File: contrib/google.golang.org/genai/chats.go (hunk 4 above)
Comment (billyac) [reply contrib/google.golang.org/genai/chats.go:1]: Moved in 0e44ff919 to contrib/google.golang.org/genai/ to match the traced import path (same as google.golang.org/grpc and google.golang.org/api).

File: contrib/googleapis/go-genai/client.go (hunk 5 above)
Comment (billyac) [reply contrib/googleapis/go-genai/client.go]: Done in 0e44ff919: added an Option type and WrapClient(c \*genai.Client, opts ...Option) — no options exported yet, but the variadic form is in place to avoid a breaking change later.

**Review by rarguelloF — APPROVED** [review]

**Comment by darccio** [comment]
@codex review

File: contrib/google.golang.org/genai/client.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+// are currently exported, but the variadic form is kept so we can add
+// them without a breaking API change.
+func WrapClient(c *genai.Client, opts ...Option) *Client {
+	if c == nil {
+		return nil
+	}
+	cfg := defaults()
+	for _, opt := range opts {
+		opt(cfg)
+	}
```
Comment (darccio) [inline contrib/google.golang.org/genai/client.go:39]: `cfg` is built and every `Option` is applied to it, but `config` currently has no fields, so this has no observable effect yet. Confirming this is the intentional placeholder from the earlier review thread (variadic `opts ...Option` kept to avoid a breaking API change later, not because an option exists today) — no action needed unless that's not the intent.

File: contrib/google.golang.org/genai/chats.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+			}
+		}
+		finishLLMSpan(span, input, c.config, acc.response(), lastErr)
+	}
+}
+
+// snapshotInput returns the history at call time plus the new user message,
+// taken before genai.Chat mutates History with the turn being sent.
+func (c *Chat) snapshotInput(parts []genai.Part) []*genai.Content {
+	src := c.chat.History(false)
```
[the remark this answers is not in the document]
Comment (darccio) [reply contrib/google.golang.org/genai/chats.go:105]: Echoing on this:

`genai.Chat.Send` builds its request from `curatedHistory` (`History(true)`), but this snapshots `History(false)` (comprehensive). When the SDK's curation drops an entry — e.g. an invalid or empty model turn — the LLMObs span's recorded `input` includes content that was never actually sent to the model, corrupting the telemetry. Snapshot the curated history instead so it matches what's sent on the wire.

```suggestion
	src := c.chat.History(true)
```

File: contrib/google.golang.org/genai/span.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+		if e == nil {
+			b.WriteString("[]")
+			continue
+		}
+		b.WriteString("[")
+		b.WriteString(itoa(len(e.Values)))
+		b.WriteString(" floats]")
+	}
+	return b.String()
+}
```
Comment (darccio) [inline contrib/google.golang.org/genai/span.go:217]: `embedSummary` renders each embedding as `"[N floats]"` rather than the actual values or any per-embedding `Statistics`/truncation info the SDK returns. That's probably the right call for span-size reasons, but confirming it's deliberate rather than an oversight — worth a short comment above `embedSummary` documenting why raw values are intentionally omitted, so a future reader doesn't "fix" it into leaking full embedding vectors into span metadata.

File: contrib/google.golang.org/genai/span.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+	var textParts []string
+	for _, p := range c.Parts {
+		if p == nil {
+			continue
+		}
+		if p.Text != "" {
+			textParts = append(textParts, p.Text)
+		}
+		if fc := p.FunctionCall; fc != nil {
+			args, _ := jsonRaw(fc.Args)
```
Comment (darccio) [inline contrib/google.golang.org/genai/span.go:149]: `args, _ := jsonRaw(fc.Args)` silently drops the error and produces a `nil` `Arguments` on failure. Since `fc.Args` is a `map[string]any` decoded from JSON, `json.Marshal` here is effectively infallible in practice, so this is low risk — but if `jsonRaw`'s error return isn't otherwise useful, consider dropping it from the helper's signature instead of swallowing it at each call site, or log via `instr` on failure so a genuinely malformed tool-call args map doesn't fail silently.

File: contrib/google.golang.org/genai/stream.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+			s.texts[idx] = &strings.Builder{}
+		}
+		// Keep the latest candidate to capture finish reason / safety ratings.
+		s.candidates[idx] = cand
+		if cand.Content != nil {
+			for _, p := range cand.Content.Parts {
+				if p != nil && p.Text != "" {
+					s.texts[idx].WriteString(p.Text)
+				}
+			}
```
Comment (darccio) [inline contrib/google.golang.org/genai/stream.go:64]: `add` only accumulates `p.Text` into a per-index `strings.Builder`; any `FunctionCall` or `FunctionResponse` part in a streamed chunk is dropped here and never reaches `response()`, whose merged `Candidate.Content.Parts` is hard-coded to a single text-only part (line 90). Non-streaming responses go through the same `contentToMessage` mapping in span.go and correctly surface `ToolCalls`/`ToolResults`, so this is a streaming-only regression: any span for a streamed tool-calling response loses its `tool_calls` entirely.

Fix needs to track non-text parts per candidate index (not just concatenated text) and re-emit them in `response()`'s merged `Content.Parts`, e.g. keeping a `[]*genai.Part` accumulator per index that appends the joined text part plus any FunctionCall/FunctionResponse parts seen, instead of only a `strings.Builder`.

**Review by darccio — CHANGES_REQUESTED** [review]
Two separate findings in this PR share the same root pattern: the wrapper's synthetic reconstruction of "what was sent/received" diverges from the SDK's actual behavior in edge cases the happy-path tests don't exercise.

- `chats.go:105` snapshots `History(false)` (comprehensive) when the SDK sends `History(true)` (curated) — diverges when a turn gets filtered.
- `stream.go:60-64` drops `FunctionCall`/`FunctionResponse` parts while accumulating streamed chunks — diverges when a streamed response contains a tool call.

Both slipped through two review rounds because the test suite has no case with a filtered history turn or a streamed tool call. Recommend fixing both together with regression tests that specifically cover those two edge cases, rather than patching the code alone — otherwise a third instance of the same class of bug is likely to ship the same way.
